autorest/responder.go

```diff
@@ -142,23 +142,34 @@ func ByUnmarshallingJSON(v interface{}) RespondDecorator {
 	}
 }
 
-// ByUnmarshallingSuccessandErrorJSON returns a RespondDecorator that decodes a JSON document returned in the
+// WithServiceErrorUnlessOK returns a RespondDecorator that decodes a JSON document returned in the
 // response Body into the value pointed by success schema v or error schema e
-func ByUnmarshallingSuccessandErrorJSON(v interface{}, e interface{}) RespondDecorator {
+func WithServiceErrorUnlessOK(v interface{}, e interface{}) RespondDecorator {
 	return func(r Responder) Responder {
 		return ResponderFunc(func(resp *http.Response) error {
 			err := r.Respond(resp)
-			b, errInner := ioutil.ReadAll(resp.Body)
-			if errInner != nil {
-				err = fmt.Errorf("Error occurred reading http.Response#Body - Error = '%v'", errInner)
-			} else if len(strings.Trim(string(b), " ")) > 0 {
-				if err == nil {
-					errInner = json.Unmarshal(b, v)
-				} else {
-					errInner = json.Unmarshal(b, e)
-				}
+			if err == nil {
+				ok := ResponseHasStatusCode(resp, http.StatusOK)
+				b, errInner := ioutil.ReadAll(resp.Body)
 				if errInner != nil {
-					err = fmt.Errorf("Error occurred unmarshalling JSON - Error = '%v' JSON = '%s'", errInner, string(b))
+					err = fmt.Errorf("Error occurred reading http.Response#Body - Error = '%v'", errInner)
+				} else if len(strings.Trim(string(b), " ")) > 0 {
+					if ok {
+						errInner = json.Unmarshal(b, v)
+					} else {
+						errInner = json.Unmarshal(b, e)
+					}
```

Comment:
@jkapil I feel this code would be simpler if it used the if-clause to select the target `interface` and then made one call to `json.Unmarshal`. Since both are interfaces, there is no need to make separate calls and the logic would then be clear (that is, it would show it is picking the target for the unmarshal).